### PR TITLE
Add `--json` output option to bundle-outdated

### PR DESCRIPTION
**Pull request by nevinera**
## What was the end-user or developer problem that led to this PR?
Building a dependency-managment nag-ops gem, we needed to parse the output of bundle-outdated. That output _is_ parseable (especially with the `--parseable` flag supplied), but I've been spoiled by tools that take the effort out of it entirely (and we see some value in exposing more information than is currently visible in that output as well).

## What is your fix for the problem, implemented in this PR?
Adding to bundle-outdated a `--json` flag, analogous to `--parseable`, but which produces json-structured data instead of human-readable line-oriented data about outdated gems.

## Make sure the following tasks are checked

- [✅] Describe the problem / feature
- [✅] Write [tests] (https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes (I could happily be more exhaustive if you like, but I tried to keep the testing comparable to the other tests on this tool)
- [✅ ] Write code to solve the problem
- [✅ ] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/lib/bundler/cli/outdated.rb
Comment (nevinera) [inline bundler/lib/bundler/cli/outdated.rb:108]: Note that this incidentally fixes a bug though probably nobody is bothered by it - in current master, if you supply both `--groups` and `--group`, the `--group` option will not be used to filter the results as expected, because of the short-circuit here.

**Comment by nevinera** [comment]
oof, lots to fix here. Back to draft!

**Review by martinemde — COMMENTED** [review]
My hesitation here is if this is the first place where we depend on json in this project.

File: bundler/lib/bundler/cli/outdated.rb
Comment (nevinera) [reply bundler/lib/bundler/cli/outdated.rb:182]: Oh, that's an excellent point! I'll update it

**Comment by nevinera** [comment]
> My hesitation here is if this is the first place where we depend on json in this project.

I've never worked in anything as broadly used or supported as bundler before - are there relevant rubies where it's not part of core? I could/should implement handling for LoadError and give a good error message if so.

Or did moving the `require` into the method address the concern?

**Comment by martinemde** [comment]
Hey @nevinera, I discussed this with @simi and I'm hoping @deivid-rodriguez will weigh in too.

Overall we think this feature seems useful. We are expecting to need json vendored in the future, but since this only requires json when the specific option is given, it's probably safe. Bundler has to be very careful about activating versions of libraries that might also be required by the bundle, which is why the require must be as narrowly scoped as possible.

If you're up for it, @simi had the excellent idea of using the user_interface elements to create a table formatter that can output to terminal format or json format. This is a large request for a first time contributor, but I figured I'd mention it if you're feeling like digging in.

Since this adds a runtime dependency, I'd like to reach a consensus of maintainers before we merge, but I'm ok with this.

**Comment by simi** [comment]
Hello @nevinera , just to add to @martinemde, my idea is to add `table` or similar method to `Bundler.ui` (which is usually `Bundler::UI::Shell` instance) and add additional `Bundler::UI::JsonShell` or something. By adding `--json` to any command, it cold just switch `Bundler.ui` and output everything in `JSON`. This new table method can be implemented in `Bundler::UI::Shell` as is printed out today and in `Bundler::UI::JsonShell` as is in your change.

Having this ready, it would be enough to generate data for the output and pas it to `Bundler.ui.table` leaving up to `Bundler.ui` decision how to "render" table.

```ruby
data = gems_list.map do |gem|
        gem_data_for(
          gem[:current_spec],
          gem[:active_spec],
          gem[:dependency],
          gem[:groups]
        )
      end

Bundler.ui.table data
```

But I'm happy to move forward on this as is, since the same parameters and output can be ported to this new UI implementation if introduced.

**Comment by nevinera** [comment]
@martinemde @simi

Ooh, I _am_ interested in that. It sounds like we could accommodate identical output and accept the same argument though, so I'd rather treat it as a next-step, if that's workable?

I'd be happy to take a swing at an implementation sometime next week - my main concern is that I'm not familiar with handling terminal size-detection; I suspect that writing a *general-purpose* `table` renderer would need some (potentially complex) logic to determine the right sizes for columns.

But if we'd just be porting out the existing implementation of the current bundle-outdated script and nothing else would initially be using that method, I'm sure I can tackle it, and leave the scary bits for someone more clever later ;-)

**Comment by martinemde** [comment]
Don't worry about making a fancy table implementation. Just porting the existing one is more than enough.

**Comment by simi** [comment]
> I suspect that writing a general-purpose table renderer would need some (potentially complex) logic to determine the right sizes for columns.

No need to print out anything fancy, current output could be considered the "table" render for non-json default UI. It would be more refactoring issue, output could be the same as today.

**Review by martinemde — APPROVED** [review]
Initial implementation looks good. Expecting to refactor in a future PR.

**Pull request merged**
